### PR TITLE
Implement per block message ids for contract level messages emitted

### PR DESCRIPTION
**Pull request by alsrdn**
This depends of https://github.com/casper-network/casper-node/pull/4461 and it replaces the per Transaction log implementation we had there. This addresses the comments from the CEP here: https://github.com/casper-network/ceps/pull/88#discussion_r1410061914 

In order to be able to track the ordering of messages that are emitted by different contracts, we now generate an unique ordering number for each message that is emitted during the execution of a block.
    
The first message that is emitted when a block is executed has `block_index=0`, the second message has `block_index=1` and so on.
We also write a record with the total number of messages that were emitted during the execution of a block. This record is purged from global state if a new message is emitted in a new block.

### CODE REVIEW

File: types/src/contract_messages/messages.rs [hunk 1]
```diff
@@ -238,6 +266,16 @@ impl Message {
...
         Key::message_topic(self.entity_addr, self.topic_name_hash)
     }
+
+    /// Returns the checksum of the message.
+    pub fn checksum(&self) -> Result<MessageChecksum, bytesrepr::Error> {
+        let mut payload_bytes = self.message.to_bytes()?;
+        let mut block_index_bytes = self.block_index.to_bytes()?;
+        block_index_bytes.append(&mut payload_bytes);
+
+        let checksum = crypto::blake2b(block_index_bytes);
```
Comment (Fraser999) [inline types/src/contract_messages/messages.rs]:
```suggestion
        let input = (&self.block_index, &self.message).to_bytes()?;
        let checksum = crypto::blake2b(input);
```

File: types/src/key.rs [hunk 2]
```diff
@@ -1358,6 +1394,7 @@ impl Distribution<Key> for Standard {
             17 => Key::AddressableEntity(rng.gen()),
             18 => Key::ByteCode(rng.gen()),
             19 => Key::Message(rng.gen()),
+            21 => Key::BlockMessageCount(BlockTime::new(rng.gen())),
```
Comment (Fraser999) [inline types/src/key.rs]: The `match` line for this needs updated (it currently has `match rng.gen_range(0..=18) {`).

File: types/src/key.rs
Comment (Fraser999) [inline types/src/key.rs:1]: We should add a `Key::BlockMessageCount` to the test collection `const KEYS`.

File: types/src/contract_messages/messages.rs (hunk 1 above)
Comment (EdHastingsCasperAssociation) [reply types/src/contract_messages/messages.rs]: @Fraser999  out of curiosity, why would you prefer to bytes over a tuple here? iirc we generally do the append style for multi value hashups.

Comment (Fraser999) [reply types/src/contract_messages/messages.rs]: Simpler code (at least IMO) but also fewer allocations 1 vs 2 or even 3 (1 per call to `to_bytes()` and possibly one in `append()`).

Comment (alsrdn) [reply types/src/contract_messages/messages.rs]: Yup, good suggestion, changed in: https://github.com/casper-network/casper-node/pull/4550/commits/4550ba389d5e0871c090f901c4c5ac5bd29648fe

File: types/src/key.rs (hunk 2 above)
Comment (alsrdn) [reply types/src/key.rs]: Added in https://github.com/casper-network/casper-node/pull/4550/commits/4550ba389d5e0871c090f901c4c5ac5bd29648fe

File: types/src/key.rs
Comment (alsrdn) [reply types/src/key.rs:1]: Added in https://github.com/casper-network/casper-node/pull/4550/commits/4550ba389d5e0871c090f901c4c5ac5bd29648fe

File: types/src/key.rs [hunk 3]
```diff
@@ -221,6 +224,8 @@ pub enum Key {
     Message(MessageAddr),
     /// A `Key` under which a single named key entry is stored.
     NamedKey(NamedKeyAddr),
+    /// A `Key` under which the total number of emitted messages in a block is stored.
+    BlockMessageCount(BlockTime),
```
Comment (EdHastingsCasperAssociation) [inline types/src/key.rs]: i have some concerns about the address-ability of this...i'd like to discuss it w/ you in realtime when we can sync up

**Review by EdHastingsCasperAssociation — APPROVED** [review]
The broad strokes LGTM. I want to deep dive into the new key variant's payload but this could merge as is and possibly get re-touched in a later PR if necessary.

**Review by Fraser999 — APPROVED** [review]

**Pull request merged**